parlai/agents/vsepp_caption/vsepp_caption.py

```diff
@@ -0,0 +1,254 @@
+        """
+        Prepares a list of candidate lists into a format ready for the model
+        as pack_padded_sequence requires each candidate must be in descending
+        order of length.
+
+        Returns a tuple of:
+        (ordered_candidate_tensor, ordered_text_candidate_list,
+         candidate_lengths, idx of truth caption*)
+        *if exists -- else it will be None
+        """
+        cand_lens = [c.shape[0] for c in candidate_vecs]
+        ind_sorted = sorted(range(len(cand_lens)), key=lambda k: -cand_lens[k])
+        truth_idx = ind_sorted.index(0) if not is_testing else None
+        cands = [candidate_labels[k] for k in ind_sorted]
+        cand_vecs = [candidate_vecs[k] for k in ind_sorted]
+        cand_lens = [cand_lens[k] for k in ind_sorted]
+        cand_lens = torch.LongTensor(cand_lens)
+
+        padded_cands = torch.LongTensor(len(candidate_vecs),
+                                        max(cand_lens)).fill_(self.NULL_IDX)
+        if self.use_cuda:
+            cand_lens = cand_lens.cuda()
+            padded_cands = padded_cands.cuda()
+
+        for i, cand in enumerate(cand_vecs):
+            padded_cands[i, :cand.shape[0]] = cand
+
+        return (padded_cands, cands, cand_lens, truth_idx)
+
+    def batch_act(self, observations):
+        batch_size = len(observations)
+        # initialize a table of replies with this agent's id
+        batch_reply = [{'id': self.getID()} for _ in range(batch_size)]
+
+        if any(['labels' in obs for obs in observations]):
+            self.mode = 'train'
+        elif any(['eval_labels' in obs for obs in observations]):
+            self.mode = 'valid'
+        else:
+            self.mode = 'test'
+
+        vec_obs = [self.vectorize(obs)
+                   for obs in observations]
+
+        # shift the labels into the text field so they're ordered
+        # by length
+        for item in observations:
+            if self.mode == 'train':
+                if 'labels' in item:
+                    item['text'] = item['labels'][0]
+                    item['text_vec'] = item['labels_vec']
+            elif self.mode == 'valid':
+                if 'eval_labels' in item:
+                    item['text'] = item['eval_labels'][0]
+                    item['text_vec'] = item['eval_labels_vec']
+
+        batch = self.batchify(vec_obs, sort=True)
+
+        images = torch.stack([self.transform(observations[idx]['image'])
+                              for idx in batch.valid_indices])
+        if self.use_cuda:
+            images = images.cuda(async=True)
+
+        if self.mode == 'train':
+            output = self.train_step(batch, images)
+        else:
+            candidate_labels = [ex['label_candidates'] for ex in vec_obs]
+            output = self.eval_step(batch, images, candidate_labels)
+
+        self.match_batch(batch_reply, batch.valid_indices, output)
+
+        return batch_reply
```

Comment:
I'd like to remove the batch_act function and just use the train_step and eval_step methods